### ISSUE TITLE
Seems there is no transform method available ?

### ISSUE THREAD

**Issue by tlunati** [YingfanWang/PaCMAP#8]
Currently working with **umap**, I can do the following:

reducer = umap.UMAP(n_neighbors=15,n_components=10)

then save the reducer using pickle

and later on, I can reuse it

reduce_data = reducer.transform(scaled_data)

PacMap does have **fit** and **fit_transform** methods but I don't find the **transform** method, so totally useless for me :-(
I can't apply an existing reducer to a new dataset.

Could you add it, I really want to benchmark it against umap (that gives me very good results)

Thanks

**Comment by robvanvolt** [comment]
> Could you try to combine the new data with the existing dataset and fit them altogether? Currently PaCMAP is designed to fit a single dataset only and cannot be applied to another dataset after fitting.

Is it a fundamental limitation, or the specific implementation of PaCMAP in this repo?

I would also be really interested in prediction, so after a PaCMAP calculation on a large dataset, where would new points be located in the previously calculated map.

**Comment by gwaygenomics** [comment]
stopping by quickly to say that I'd be interested in this functionality as well. Thanks for putting this repo out there in the first place @YingfanWang!

**Comment by candalfigomoro** [comment]
> Could you try to combine the new data with the existing dataset and fit them altogether? Currently PaCMAP is designed to fit a single dataset only and cannot be applied to another dataset after fitting.

@hyhuang00 
For example, you could use PaCMAP as a preprocessing step for a supervised learning problem. In a supervised learning problem, you typically split data in training/(validation)/test sets. The problem is that, if you combine all the data and fit a single dataset, you are leaking information across datasets because the representation for training samples also depends on test samples. So, we would need to fit just the training set and then project/transform new points (test points) using the previously calculated map. In this way, we are not leaking information from the test set into the training set.

As a workaround (this is something I tried also for UMAP before it provided a transform method), you could try to train a model (e.g. a simple neural network) to learn the mapping from the original space to the "PaCMAPped" space, then use this model to transform new data. I can't say if it will work well, but this is something you can try, especially if you have enough data to train the model.

**Comment by WiestDaessle** [comment]
@gwaygenomics : Just noticed you participated here ! My update may suit your needs, was doing a test on our JUMP-CP dataset...

**Comment by gwaygenomics** [comment]
Thanks! Excited to explore once it's reviewed!

**Comment by hyhuang00** [comment]
@candalfigomoro Apologies for my late reply! I'm so sorry I have missed your suggestions. We have been working on this for a while but we haven't finished a good implementation of this yet. Fortunately @WiestDaessle has provided a great contribution on this direction and we will accept the pull request once the code has been reviewed. Thank you all for your interest and contribution!

**Comment by hyhuang00** [comment]
The transform method has been implemented in the newest release 0.6.0. Transform method will create nearest neighbor pairs between existing dataset and the new dataset. A demo and a short tutorial will be added shortly!

**Comment by nfultz** [comment]
@hyhuang00 that is great, thanks much.

When I implemented my workaround (https://github.com/njnmco/smc/blob/main/src/common/reduction.py#L26) I noticed that successive calls to `pacmap() ` could return arbitrary rotations, so I used OLS to rotate them back to the original fitted directions. Will a similar workaround be needed for the new `transform()` feature?
